docs/design/generic-enhanced-purls.md

```diff
@@ -0,0 +1,286 @@
+
+There are use cases that would benefit from more accurate purls, though, such as the recent Maven artifacts [proposal](https://github.com/containerbuildsystem/cachi2/pull/663). Considering that the purl specification already identifies several types of packages that don't fit into traditional package manager (e.g. github, docker, huggingface; see the [purl types spec](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst) for more info), this proposal builds on top of the fundamentals of the generic fetcher to provide an extensible mechanism that would allow Cachi2 to fetch files from specific sources and report them with matching purl types.
+
+## Consuming input purls
+
+A purl is simply a way to represent a unique package and its location on the internet. By allowing users to specify a purl as part of a package definition in the lockfile, Cachi2 can verify that it actually resolves to a valid download location, and that it adheres to the fundamentals of the purl spec. Note that there's a great variety of purl types, and resolving them will involve type-specific logic, which means we'll only be able to support a small subset of purl types.
+
+This is a summary of how the resolution of a single package will look like:
+
+- Instead of consuming a `download_url`, the package will specify a `purl` attribute in the lockfile
+- Parse the purl by using the [packageurl-python](https://github.com/package-url/packageurl-python) library
+- Validate that the purl is within the supported types
+- Validate that the qualifiers are valid for the specific type (see the [type definitions](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst) for type-specific qualifiers and also the list of [type-agnostic qualifiers](https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#known-qualifiers-keyvalue-pairs))
+- Resolve the purl to a download URL by using a type-specific algorithm
+- Download the file
+- In case the artifact is a file, verify at least one of the provided checksums match
+- Generate a SBOM component containing the input purl with a single verified checksum added to it (this is to avoid having a large number of checksums cluttering the purl)
+
+### What confidence Cachi2 has in the reported SBOM components
+
+- The purl is syntatically valid
+- The purl resolves to an actual URL where the artifact was fetched from
+- The artifact's checksum was validated
+- There is no guarantee over the reported purl type, Cachi2 will trust the user input
+
+## Examples of workflows for different purl types
+
+### Maven
+
+#### generic_artifacts.yaml
+
+```yaml
+metadata:
+  version: 1.0.0
+artifacts:
+  - purl: pkg:maven/io.quarkus/quarkus-core@3.8.5.redhat-00004?type=jar&repository_url=https://maven.repository.redhat.com/ga
+    checksums:
+      sha256: d16bf783cb6670f7f692ad7d6885ab957c63cfc1b9649bc4a3ba1cfbdfd5230c
+    target: quarkus.jar
```

Comment:
Is this target the location on-disk for where the fetch artifacts would be put? Would all artifacts need to specify this so that they can be appropriately used by some later stage or would there be default/assumptions for package managers about locations and file names?